src/SDK/SmartDetectorsShared/Tools/ObjectPool.cs

```
@@ -0,0 +1,101 @@
+// </copyright>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.Azure.Monitoring.SmartDetectors.Tools
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+
+    /// <summary>
+    /// Pool of objects that can be leased.
+    /// </summary>
+    /// <typeparam name="T">Type of objects in pool</typeparam>
+    public class ObjectPool<T>
+    {
+        private readonly int maxPoolSize;
+        private readonly Func<T> objectGenerator;
+        private readonly BlockingCollection<T> pool = new BlockingCollection<T>();
+        
+        private long poolSize = 0;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObjectPool{T}"/> class.
+        /// </summary>
+        /// <param name="objectGenerator">Function that generates a new object in the case the pool is empty</param>
+        /// <param name="maxPoolSize">Maximum amount of objects in pool.  Default is 100.</param>
+        public ObjectPool(Func<T> objectGenerator, int maxPoolSize = 100)
+        {
+            this.objectGenerator = objectGenerator;
+            this.maxPoolSize = maxPoolSize;
+        }
+
+        /// <summary>
+        /// Returns an item from the pool.  If no items exist in the pool either:
+        /// 1. A new item is created if the number of items created is less than the max.
+        /// 2. Blocks until an item becomes available.
+        /// </summary>
+        /// <returns>Item from the pool</returns>
+        public ILeasedItem<T> LeaseItem()
+        {
+            if (this.pool.TryTake(out T retval))
+            {
+                return new LeasedItem<T>(this, retval);
+            }
+
+            if (Interlocked.Increment(ref this.poolSize) > this.maxPoolSize)
+            {
+                Interlocked.Decrement(ref this.poolSize);
+                return new LeasedItem<T>(this, this.pool.Take());

```

>Take [](start = 57, length = 4)

not too keen about this being a blocking operation - can you add an async semaphore and signal on it when you release an item?

---

I can do it, but it will be a flavor of the method that won't be used anywhere (unless you think we should also change Hash() to an async method).  I think it makes more sense to add this when we actually need so we don't have code around that is not actually used.

Another option is to increase the pool size and in the case that no object is available throw and fail fast.  What do you think?

---
In reply to: [225815703](https://github.com/liadtal/Smart-Alerts/pull/230#discussion_r225815703) [](ancestors = 225815703)